C7/Map/MapView.cs

```diff
@@ -0,0 +1,445 @@
+		private string[,]terrain;
+		private TileMap terrainTilemap;
+		private TileMap terrainTilemapShadow;
+		private TileSet terrainTileset;
+		private TileMap tilemap;
+		private TileSet tileset;
+		private Vector2I tileSize = new Vector2I(128, 64);
+		private ILogger log = LogManager.ForContext<MapView>();
+		public int worldEdgeRight {get; private set;}
+		public int worldEdgeLeft {get; private set;}
+		private int width;
+		private int height;
+		public bool showGrid {
+			get => showGrid;
+			set {
+				showGrid = value;
+			}
+		}
+		private Game game;
+		private GameMap gameMap;
+
+		public override void _Draw() {
+			GD.Print("draw...");
+			game.animTracker.update();
+			foreach ((string id, AnimationTracker.ActiveAnimation anim) in game.animTracker.activeAnims) {
+				GD.Print($"{id}: {anim.ToString()}");
+			}
+			base._Draw();
+		}
+
+		public override void _Process(double delta) {
+			base._Process(delta);
+		}
+
+		private void initializeTileMap() {
+			terrainTilemap = new TileMap();
+			terrainTilemapShadow = new TileMap();
+			terrainTileset = Civ3TerrainTileSet.Generate();
+			terrainTilemap.TileSet = terrainTileset;
+			terrainTilemap.Position += Vector2I.Right * (tileSize.X / 2);
+			terrainTilemapShadow.TileSet = terrainTileset;
+			terrainTilemapShadow.Position = terrainTilemap.Position + (Vector2I.Left * tileSize.X * width);
+
+			tilemap = new TileMap{ YSortEnabled = true };
+			tileset = TileSetLoader.LoadCiv3TileSet();
+			tilemap.TileSet = tileset;
+
+			// create tilemap layers
+			foreach (Layer layer in Enum.GetValues(typeof(Layer))) {
+				if (layer != Layer.Invalid) {
+					tilemap.AddLayer(layer.Index());
+					tilemap.SetLayerYSortEnabled(layer.Index(), true);
+				}
+			}
+
+			tilemap.ZIndex = 10; // need to figure out a good way to order z indices
+			AddChild(tilemap);
+			AddChild(terrainTilemap);
+			// AddChild(terrainTilemapShadow);
+		}
+
+		private void setTerrainTiles() {
+			string[] corners(int x, int y) {
+				string left = terrain[x, y];
+				string right = terrain[(x + 1) % width, y];
+				bool even = y % 2 == 0;
+				string top = "coast";
+				if (y > 0) {
+					top = even ? terrain[x, y - 1] : terrain[(x + 1) % width, y - 1];
+				}
+				string bottom = "coast";
+				if (y < height - 1) {
+					bottom = even ? terrain[x, y + 1] : terrain[(x + 1) % width, y + 1];
+				}
+				return new string[4]{top, right, bottom, left};
+			}
+			void lookupAndSetTerrainTile(int x, int y, int cellX, int cellY) {
+				Vector2I cell = new Vector2I(cellX, cellY);
+				string[] corner = corners(x, y);
+				TerrainPcx pcx = Civ3TerrainTileSet.GetPcxFor(corner);
+				Vector2I texCoords = pcx.getTextureCoords(corner);
+				setTerrainTile(cell, pcx.atlas, texCoords);
+			}
+			for (int x = 0; x < width; x++) {
+				for (int y = 0; y < height; y++) {
+					lookupAndSetTerrainTile(x, y, x, y);
+				}
+			}
+			for (int y = 0; y < height; y++) {
+				lookupAndSetTerrainTile(width - 1, y, -1, y);
+			}
```

Comment:
only because I've seen several of these "do for all values in range" blocks, there are more concise ways to do this with LINQ using SelectMany, but I'm not sure it provides any real advantage

Reply:
Not familiar with SelectMany so I read the documentation:
Projects each element of a sequence to an [IEnumerable<T>](https://learn.microsoft.com/en-us/dotnet/api/system.collections.generic.ienumerable-1?view=net-7.0), flattens the resulting sequences into one sequence, and invokes a result selector function on each element therein.

For me at least, having to read and understand this is much more difficult than reading the more verbose for loops. Perhaps the linq query can generate faster code so maybe there's a tradeoff there.